src/SchemaMigrations/DefinedSchemas.js

```diff
@@ -0,0 +1,417 @@
+        await this.wait(1000 * this.retries);
+        await this.executeMigrations();
+      } else {
+        logger.error(`Failed to run migrations: ${e}`);
+        if (process.env.NODE_ENV === 'production') process.exit(1);
+      }
+    }
+  }
+
+  checkForMissingSchemas() {
+    if (this.schemaOptions.strict !== true) {
+      return;
+    }
+
+    const cloudSchemas = this.allCloudSchemas.map(s => s.className);
+    const localSchemas = this.localSchemas.map(s => s.className);
+    const missingSchemas = cloudSchemas.filter(
+      c => !localSchemas.includes(c) && !systemClasses.includes(c)
+    );
+
+    if (new Set(localSchemas).size !== localSchemas.length) {
+      logger.error(
+        `The list of schemas provided contains duplicated "className"  "${localSchemas.join(
+          '","'
+        )}"`
+      );
+      process.exit(1);
+    }
+
+    if (this.schemaOptions.strict && missingSchemas.length) {
+      logger.warn(
+        `The following schemas are currently present in the database, but not explicitly defined in a schema: "${missingSchemas.join(
+          '", "'
+        )}"`
+      );
+    }
+  }
+
+  // Required for testing purpose
+  async wait(time) {
+    await new Promise(resolve => setTimeout(resolve, time));
+  }
+
+  async enforceCLPForNonProvidedClass(): void {
+    const nonProvidedClasses = this.allCloudSchemas.filter(
+      cloudSchema =>
+        !this.localSchemas.some(localSchema => localSchema.className === cloudSchema.className)
+    );
+    await Promise.all(
+      nonProvidedClasses.map(async schema => {
+        const parseSchema = new Parse.Schema(schema.className);
+        this.handleCLP(schema, parseSchema);
+        await this.updateSchemaToDB(parseSchema);
+      })
+    );
+  }
+
+  // Create a fake session since Parse do not create the _Session until
+  // a session is created
+  async createDeleteSession() {
+    const session = new Parse.Session();
+    await session.save(null, { useMasterKey: true });
+    await session.destroy({ useMasterKey: true });
+  }
+
+  async saveOrUpdate(localSchema: Migrations.JSONSchema) {
+    const cloudSchema = this.allCloudSchemas.find(sc => sc.className === localSchema.className);
+    if (cloudSchema) {
+      try {
+        await this.updateSchema(localSchema, cloudSchema);
+      } catch (e) {
+        throw `Error during update of schema for type ${cloudSchema.className}: ${e}`;
+      }
+    } else {
+      try {
+        await this.saveSchema(localSchema);
+      } catch (e) {
+        throw `Error while saving Schema for type ${localSchema.className}: ${e}`;
+      }
+    }
+  }
+
+  async saveSchema(localSchema: Migrations.JSONSchema) {
+    const newLocalSchema = new Parse.Schema(localSchema.className);
+    if (localSchema.fields) {
+      // Handle fields
+      Object.keys(localSchema.fields)
+        .filter(fieldName => !this.isProtectedFields(localSchema.className, fieldName))
+        .forEach(fieldName => {
```

Comment:
I don't like this syntax personally, I think it's a lot more readable to do something such as:

```js
const fields = Object.keys(localSchema.fields);
const filteredFields = fields.filter(fieldName => !this.isProtectedFields(localSchema.className, fieldName))
for (const fieldName of filteredFields) {
  const field = localSchema.fields[fieldName];
  this.handleFields(newLocalSchema, fieldName, field);
}
```

Reply:
hummmm here from my point of view we should stick to pipe functional flow style 🙂  not a huge fan of "for" loops

Reply:
`for of` loops are great! No worries, I personally find it more readable + it allows for control flow (`continue` `break`), but that's just my opinion 😊

Reply:
@mtrezza do you have any thoughts on preferred syntax